Notebook modified from https://github.com/data-8/textbook

In [1]:
# from datascience import *
import matplotlib
path_data = '../../../assets/data/'
import pandas as pd
pd.set_option('display.max_rows', 10)

# Example: Population Trends

We are now ready to work with large tables of data. The file below contains "Annual Estimates of the Resident Population by Single Year of Age and Sex for the United States." Notice that `read_table` can read data directly from a URL.

In [10]:
# As of Jan 2017, this census file is online here: 
data = 'http://www2.census.gov/programs-surveys/popest/datasets/2010-2015/national/asrh/nc-est2015-agesex-res.csv'

# A local copy can be accessed here in case census.gov moves the file:
# data = path_data + 'nc-est2015-agesex-res.csv'

full_census_table = pd.read_csv(data)
full_census_table

SEX  AGE  CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  \
0      0    0        3944153            3944160          3951330   
1      0    1        3978070            3978090          3957888   
2      0    2        4096929            4096939          4090862   
3      0    3        4119040            4119051          4111920   
4      0    4        4063170            4063186          4077551   
..   ...  ...            ...                ...              ...   
301    2   97          53582              53605            54118   
302    2   98          36641              36675            37532   
303    2   99          26193              26214            26074   
304    2  100          44202              44246            45058   
305    2  999      156964212          156969328        157258820   

     POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  POPESTIMATE2014  \
0            3963087          3926540          3931141          3949775   
1            3966551          3977939          3942872          3949776   
2            3971565          3980095          3992720          3959664   
3            4102470          3983157          3992734          4007079   
4            4122294          4112849          3994449          4005716   
..               ...              ...              ...              ...   
301            57159            59533            61255            62779   
302            40116            42857            44359            46208   
303            27030            29320            31112            32517   
304            47556            50661            53902            58008   
305        158427085        159581546        160720625        161952064   

     POPESTIMATE2015  
0            3978038  
1            3968564  
2            3966583  
3            3974061  
4            4020035  
..               ...  
301            69285  
302            47272  
303            34064  
304            61886  
305        163189523  

[306 rows x 10 columns]

Only the first 10 rows of the table are displayed. Later we will see how to display the entire table; however, this is typically not useful with large tables.

a [description of the table](http://www2.census.gov/programs-surveys/popest/datasets/2010-2015/national/asrh/nc-est2015-agesex-res.pdf) appears online. The `SEX` column contains numeric codes: `0` stands for the total, `1` for male, and `2` for female. The `AGE` column contains ages in completed years, but the special value `999` is a sum of the total population. The rest of the columns contain estimates of the US population.

Typically, a public table will contain more information than necessary for a particular investigation or analysis. In this case, let us suppose that we are only interested in the population changes from 2010 to 2014. Let us select the relevant columns using `[]`.

In [11]:
partial_census_table = full_census_table[['SEX', 'AGE', 'POPESTIMATE2010', 'POPESTIMATE2014']]
partial_census_table

SEX  AGE  POPESTIMATE2010  POPESTIMATE2014
0      0    0          3951330          3949775
1      0    1          3957888          3949776
2      0    2          4090862          3959664
3      0    3          4111920          4007079
4      0    4          4077551          4005716
..   ...  ...              ...              ...
301    2   97            54118            62779
302    2   98            37532            46208
303    2   99            26074            32517
304    2  100            45058            58008
305    2  999        157258820        161952064

[306 rows x 4 columns]

We can also simplify the labels of the selected columns.

In [12]:
us_pop = partial_census_table.rename(columns={'POPESTIMATE2010': '2010', 'POPESTIMATE2014': '2014'})
us_pop

SEX  AGE       2010       2014
0      0    0    3951330    3949775
1      0    1    3957888    3949776
2      0    2    4090862    3959664
3      0    3    4111920    4007079
4      0    4    4077551    4005716
..   ...  ...        ...        ...
301    2   97      54118      62779
302    2   98      37532      46208
303    2   99      26074      32517
304    2  100      45058      58008
305    2  999  157258820  161952064

[306 rows x 4 columns]

We now have a table that is easy to work with. Each column of the table is an array of the same length, and so columns can be combined using arithmetic. Here is the change in population between 2010 and 2014.

In [13]:
us_pop['2014'] - us_pop['2010']

0        -1555
1        -8112
2      -131198
3      -104841
4       -71835
        ...   
301       8661
302       8676
303       6443
304      12950
305    4693244
Length: 306, dtype: int64

Let us augment `us_pop` with a column that contains these changes, both in absolute terms and as percents relative to the value in 2010.

In [15]:
census = pd.DataFrame(us_pop)

us_pop['Change'] = us_pop['2014'] - us_pop['2010']
us_pop['Percent Change'] = 100 * us_pop['Change'] / us_pop['2010']

census

SEX  AGE       2010       2014   Change  Percent Change
0      0    0    3951330    3949775    -1555       -0.039354
1      0    1    3957888    3949776    -8112       -0.204958
2      0    2    4090862    3959664  -131198       -3.207099
3      0    3    4111920    4007079  -104841       -2.549685
4      0    4    4077551    4005716   -71835       -1.761719
..   ...  ...        ...        ...      ...             ...
301    2   97      54118      62779     8661       16.003917
302    2   98      37532      46208     8676       23.116274
303    2   99      26074      32517     6443       24.710440
304    2  100      45058      58008    12950       28.740734
305    2  999  157258820  161952064  4693244        2.984407

[306 rows x 6 columns]

**Sorting the data.** Let us sort the table in decreasing order of the absolute change in population.

In [18]:
census.sort_values('Change', ascending=False)

SEX  AGE       2010       2014   Change  Percent Change
101    0  999  309346863  318907401  9560538        3.090556
203    1  999  152088043  156955337  4867294        3.200313
305    2  999  157258820  161952064  4693244        2.984407
67     0   67    2693707    3485241   791534       29.384562
64     0   64    2706055    3487559   781504       28.879827
..   ...  ...        ...        ...      ...             ...
39     0   39    4324463    3982507  -341956       -7.907479
48     0   48    4534663    4159738  -374925       -8.267979
46     0   46    4529716    4077151  -452565       -9.991024
47     0   47    4535473    4082883  -452590       -9.978893
40     0   40    4387480    3859395  -528085      -12.036180

[306 rows x 6 columns]

Not surprisingly, the top row of the sorted table is the line that corresponds to the entire population: both sexes and all age groups. From 2010 to 2014, the population of the United States increased by about 9.5 million people, a change of just over 3%.

The next two rows correspond to all the men and all the women respectively. The male population grew more than the female population, both in absolute and percentage terms. Both percent changes were around 3%.

Now take a look at the next few rows. The percent change jumps from about 3% for the overall population to almost 30% for the people in their late sixties and early seventies. This stunning change contributes to what is known as the greying of America.

By far the greatest absolute change was among those in the 64-67 agegroup in 2014. What could explain this large increase? We can explore this question by examining the years in which the relevant groups were born.

- Those who were in the 64-67 age group in 2010 were born in the years 1943 to 1946. The attack on Pearl Harbor was in late 1941, and by 1942 U.S. forces were heavily engaged in a massive war that ended in 1945. 

- Those who were 64 to 67 years old in 2014 were born in the years 1947 to 1950, at the height of the post-WWII baby boom in the United States. 

The post-war jump in births is the major reason for the large changes that we have observed.